docs/source/for-ops/configuration.rst
.. _configuration:

Configuration
#############

Circus can be configured using an ini-style configuration file.

Example:

.. code-block:: ini

    [cvec_circus]
    check_delay = 5
    endpoint = tcp://127.0.0.1:5555
    pubsub_endpoint = tcp://127.0.0.1:5556
    include = \*.more.config.ini
    umask = 002

    [watcher:myprogram]
    cmd = python
    args = -u myprogram.py $(cvec_circus.wid) $(CIRCUS.ENV.VAR)
    warmup_delay = 0
    numprocesses = 5

    # hook
    hooks.before_start = my.hooks.control_redis

    # will push in test.log the stream every 300 ms
    stdout_stream.class = FileStream
    stdout_stream.filename = test.log

    # optionally rotate the log file when it reaches 1 gb
    # and save 5 copied of rotated files
    stdout_stream.max_bytes = 1073741824
    stdout_stream.backup_count = 5

    [env:myprogram]
    PATH = $PATH:/bin
    CAKE = lie

    [plugin:statsd]
    use = cvec_circus.plugins.statsd.StatsdEmitter
    host = localhost
    port = 8125
    sample_rate = 1.0
    application_name = example

    [socket:web]
    host = localhost
    port = 8080


cvec_circus - single section
=======================
    **endpoint**
        The ZMQ socket used to manage Circus via **cvec_circusctl**.
        (default: *tcp://127.0.0.1:5555*)
    **endpoint_owner**
        If set to a system username and the endpoint is an ipc socket like
        *ipc://var/run/cvec_circusd.sock*, then ownership of the socket file will
        be changed to that user at startup. For more details, see :ref:`security`.
        (default: None)
    **pubsub_endpoint**
        The ZMQ PUB/SUB socket receiving publications of events.
        (default: *tcp://127.0.0.1:5556*)
    **papa_endpoint**
        If using :ref:`papa`, you can specify the endpoint, such as
        *ipc://var/run/cvec_circusd.sock*.
        (default: *tcp://127.0.0.1:20202*)
    **statsd**
        If set to True, Circus runs the cvec_circusd-stats daemon. (default: False)
    **stats_endpoint**
        The ZMQ PUB/SUB socket receiving publications of stats.
        (default: *tcp://127.0.0.1:5557*)
    **statsd_close_outputs**
        If True sends the cvec_circusd-stats stdout/stderr to ``/dev/null``.
        (default: False)
    **check_delay**
        The polling interval in seconds for the ZMQ socket. (default: 5)
    **include**
        List of config files to include. You can use wildcards
        (`*`) to include particular schemes for your files. The paths are
        absolute or relative to the config file. (default: None)
    **include_dir**
        List of config directories. All files matching `*.ini` under each
        directory will be included. The paths are absolute or relative to the
        config file. (default: None)
    **stream_backend**
        Defines the type of backend to use for the streaming. Possible
        values are **thread** or **gevent**. (default: thread)
    **warmup_delay**
        The interval in seconds between two watchers start. Must be an int. (default: 0)
    **httpd**
        If set to True, Circus runs the cvec_circushttpd daemon. (default: False)
    **httpd_host**
        The host ran by the cvec_circushttpd daemon. (default: localhost)
    **httpd_port**
        The port ran by the cvec_circushttpd daemon. (default: 8080)
    **httpd_close_outputs**
        If True, sends the cvec_circushttpd stdout/stderr to ``/dev/null``.
        (default: False)
    **debug**
        If set to True, all Circus stout/stderr daemons are redirected to cvec_circusd
        stdout/stderr (default: False)
    **debug_gc**
        If set to True, cvec_circusd outputs additional log info from the garbage
        collector. This can be useful in tracking down memory leaks.
        (default: False)
    **pidfile**
        The file that must be used to keep the daemon pid.
    **umask**
        Value for umask. If not set, cvec_circusd will not attempt to modify umask.
    **loglevel**
        The loglevel that we want to see (default: INFO)
    **logoutput**
        The logoutput file where we want to log (default: ``-`` to log
        on stdout). You can log to a remote syslog by using the
        following syntax: ``syslog://host:port?facility`` where host
        is your syslog server, port is optional and facility is the
        syslog facility to use. If you wish to log to a local syslog
        you can use ``syslog:///path/to/syslog/socket?facility``
        instead.
    **loggerconfig**
        A path to an INI, JSON or YAML file to configure standard Python
        logging for the Arbiter.  The special value "default" uses the builtin
        logging configuration based on the optional loglevel and logoutput options.
	
        **Example YAML Configuration File**

    .. code-block:: yaml

            version: 1
            disable_existing_loggers: false
            formatters:
              simple:
                format: '%(asctime)s - %(name)s - [%(levelname)s] %(message)s'
            handlers:
              logfile:
                class: logging.FileHandler
                filename: logoutput.txt
                level: DEBUG
                formatter: simple
            loggers:
              cvec_circus:
                level: DEBUG
                handlers: [logfile]
                propagate: no
            root:
              level: DEBUG
              handlers: [logfile]

watcher:NAME - as many sections as you want
===========================================
    **NAME**
        The name of the watcher. This name is used in **cvec_circusctl**
    **cmd**
        The executable program to run.
    **args**
        Command-line arguments to pass to the program. You can use the python
        format syntax here to build the parameters. Environment variables are
        available, as well as the worker id and the environment variables that
        you passed, if any, with the "env" parameter. See
        :ref:`formatting_cmd` for more information on this.
    **shell**
        If True, the processes are run in the shell (default: False)
    **shell_args**
        Command-line arguments to pass to the shell command when **shell** is
        True. Works only for \*nix system (default: None)
    **working_dir**
        The working dir for the processes (default: None)
    **uid**
        The user id or name the command should run with.
        (The current uid is the default).
    **gid**
        The group id or name the command should run
        with. (The current gid is the default).
    **copy_env**
        If set to true, the local environment variables will be copied and
        passed to the workers when spawning them. (Default: False)
    **copy_path**
        If set to true, **sys.path** is passed in the subprocess environ
        using *PYTHONPATH*. **copy_env** has to be true.
        (Default: False)
    **warmup_delay**
        The delay (in seconds) between running processes.
    **autostart**
        If set to false, the watcher will not be started automatically
        when the arbiter starts. The watcher can be started explicitly
        (example: `cvec_circusctrl start myprogram`). (Default: True)
    **numprocesses**
        The number of processes to run for this watcher.
    **rlimit_LIMIT**
        Set resource limit LIMIT for the watched processes. The
        config name should match the RLIMIT_* constants (not case
        sensitive) listed in the `Python resource module reference
        <http://docs.python.org/library/resource.html#resource-limits>`_.
        For example, the config line 'rlimit_nofile = 500' sets the maximum
        number of open files to 500. To set a limit value to RLIM_INFINITY,
        do not set a value, like this config line: 'rlimit_nofile = '.
    **stderr_stream.class**
        A fully qualified Python class name that will be instanciated, and
        will receive the **stderr** stream of all processes in its
        :func:`__call__` method.

        Circus provides some stream classes you can use without prefix:

        - :class:`FileStream`: writes in a file and can do automatic log rotation
        - :class:`WatchedFileStream`: writes in a file and relies on external log rotation
        - :class:`TimedRotatingFileStream`: writes in a file and can do rotate at certain timed intervals.
        - :class:`QueueStream`: write in a memory Queue
        - :class:`StdoutStream`: writes in the stdout
        - :class:`FancyStdoutStream`: writes colored output with time prefixes in the stdout

    **stderr_stream.***
        All options starting with *stderr_stream.* other than *class* will
        be passed the constructor when creating an instance of the
        class defined in **stderr_stream.class**.
    **stdout_stream.class**
        A fully qualified Python class name that will be instanciated, and
        will receive the **stdout** stream of all processes in its
        :func:`__call__` method.

        Circus provides some stream classes you can use without prefix:

        - :class:`FileStream`: writes in a file and can do automatic log rotation
        - :class:`WatchedFileStream`: writes in a file and relies on external log rotation
        - :class:`TimedRotatingFileStream`: writes in a file and can do rotate at certain timed intervals.
        - :class:`QueueStream`: write in a memory Queue
        - :class:`StdoutStream`: writes in the stdout
        - :class:`FancyStdoutStream`: writes colored output with time prefixes in the stdout

    **stdout_stream.***
        All options starting with *stdout_stream.* other than *class* will
        be passed the constructor when creating an instance of the
        class defined in **stdout_stream.class**.

    **stdin_socket**
        If not None, the socket with matching name is placed at file descriptor 0 (stdin)
	of the processes. (Default: None)

    **close_child_stdin**
        If set to True, the stdin stream of each process will be sent to
        ``/dev/null`` after the fork. Defaults to True.

        The primary use case for this option is debugging. Set it to False and
        Circus will leave *stdin* open for the watcher processes. This allows
        interactive debugger sessions to be attached to them. In case of a
        Python program, one could insert a Pdb call (``import pdb;
        pdb.set_trace()``) somewhere in the code and acquire its shell to
        debug the program.

        If debugging a process is not expected, leave this option in its default
        value, True. This will prevent, for example, that the process hangs
        indefinitely waiting on input which could never come, especially if
        Circus runs as a daemon.

    **close_child_stdout**
        If set to True, the stdout stream of each process will be sent to
        ``/dev/null`` after the fork. Defaults to False.

    **close_child_stderr**
        If set to True, the stderr stream of each process will be sent to
        ``/dev/null`` after the fork. Defaults to False.

    **send_hup**
        If True, a process reload will be done by sending the SIGHUP signal.
        Defaults to False.

    **stop_signal**
        The signal to send when stopping the process. Can be specified as a
        number or a signal name. Signal names are case-insensitive and can
        include 'SIG' or not. So valid examples include `quit`, `INT`,
        `SIGTERM` and `3`.
        Defaults to SIGTERM.

    **stop_children**
        When sending the *stop_signal*, send it to the children as well.
        Defaults to False.

    **max_retry**
        The number of times we attempt to start a process, before
        we abandon and stop the whole watcher. Defaults to 5.
        Set to -1 to disable max_retry and retry indefinitely.

.. _graceful_timeout:

    **graceful_timeout**
        The number of seconds to wait for a process to terminate gracefully
        before killing it.

        When stopping a process, we first send it a *stop_signal*. A worker
        may catch this signal to perform clean up operations before exiting.
        If the worker is still active after graceful_timeout seconds, we send
        it a SIGKILL signal.  It is not possible to catch SIGKILL signals so
        the worker will stop.

        Defaults to 30s.

    **priority**
        Integer that defines a priority for the watcher. When the
        Arbiter do some operations on all watchers, it will sort them
        with this field, from the bigger number to the smallest.
        Defaults to 0.

    **singleton**
        If set to True, this watcher will have at the most one process.
        Defaults to False.

    **use_sockets**
        If set to True, this watcher will be able to access defined sockets
        via their file descriptors. If False, all parent fds are closed
        when the child process is forked. Defaults to False.

    **max_age**
        If set then the process will be restarted sometime after max_age
        seconds. This is useful when processes deal with pool of connectors:
        restarting processes improves the load balancing. Defaults to being
        disabled.

    **max_age_variance**
        If max_age is set then the process will live between max_age and
        max_age + random(0, max_age_variance) seconds. This avoids restarting
        all processes for a watcher at once. Defaults to 30 seconds.

    **on_demand**
        If set to True, the processes will be started only after the first
        connection to one of the configured sockets (see below). If a restart
        is needed, it will be only triggered at the next socket event.

    **hooks.***
        Available hooks: **before_start**, **after_start**,
        **before_spawn**, **after_spawn**,
        **before_stop**, **after_stop**,
        **before_signal**, **after_signal**,
        **extended_stats**

        Define callback functions that hook into the watcher startup/shutdown process.

        If the hook returns **False** and if the hook is one of
        **before_start**, **before_spawn**, **after_start** or **after_spawn**,
        the startup will be aborted.

        If the hook is **before_signal** and returns **False**, then the
        corresponding signal will not be sent (except SIGKILL which is always
        sent)

        Notice that a hook that fails during the stopping process will not
        abort it.

        The callback definition can be followed by a boolean flag separated by a
        comma. When the flag is set to **true**, any error occuring in the
        hook will be ignored. If set to **false** (the default), the hook
        will return **False**.

        More on :ref:`hooks`.

    **virtualenv**
        When provided, points to the root of a Virtualenv directory. The
        watcher will scan the local **site-packages** and loads its content
        into the execution environment. Must be used with **copy_env** set
        to True. Defaults to None.

    **virtualenv_py_ver**
        Specifies the python version of the virtualenv (e.g "3.3").
        It's usefull if cvec_circus run with another python version (e.g "2.7")
        The watcher will scan the local **site-packages** of the specified
        python version and load its content into the execution
        environment. Must be used with **virtualenv**. Defaults to None.

    **respawn**
        If set to False, the processes handled by a watcher will not be
        respawned automatically. The processes can be manually respawned with
        the `start` command. (default: True)

    **use_papa**
        Set to true to use the :ref:`papa`.



socket:NAME - as many sections as you want
==========================================
    **host**
        The host of the socket. Defaults to 'localhost'
    **port**
        The port. Defaults to 8080.
    **family**
        The socket family. Can be 'AF_UNIX', 'AF_INET' or 'AF_INET6'.
        Defaults to 'AF_INET'.
    **type**
        The socket type. Can be 'SOCK_STREAM', 'SOCK_DGRAM', 'SOCK_RAW',
        'SOCK_RDM' or 'SOCK_SEQPACKET'. Defaults to 'SOCK_STREAM'.
    **interface**
        When provided a network interface name like 'eth0', binds the socket
        to that particular device so that only packets received from that
        particular interface are processed by the socket.
        This can be used for example to limit which device to bind when
        binding on IN_ADDR_ANY (0.0.0.0) or IN_ADDR_BROADCAST
        (255.255.255.255). Note that this only works for some socket types,
        particularly AF_INET sockets.
    **path**
        When provided a path to a file that will be used as a unix socket
        file. If a path is provided, **family** is forced to AF_UNIX and
        **host** and **port** are ignored.
    **umask**
        When provided, sets the umask that will be used to create an
        AF_UNIX socket. For example, `umask=000` will produce a socket with
        permission `777`.
    **replace**
        When creating Unix sockets ('AF_UNIX'), an existing file may indicate a
        problem so the default is to fail. Specify `True` to simply remove the
        old file if you are sure that the socket is managed only by Circus.
    **so_reuseport**
        If set to True and SO_REUSEPORT is available on target platform, cvec_circus
        will create and bind new SO_REUSEPORT socket(s) for every worker it starts
        which is a user of this socket(s).
    **blocking**
        If `True`, socket is set to blocking. If `False`, socket is set to non-blocking.
        (default: False)

    **use_papa**
        Set to true to use the :ref:`papa`.


Once a socket is created, the *${cvec_circus.sockets.NAME}* string can be used in the
command (`cmd` or `args`) of a watcher. Circus will replace it by the FD value. The watcher must also
have `use_sockets` set to `True` otherwise the socket will have been closed and
you will get errors when the watcher tries to use it.

Example:

.. code-block:: ini

    [watcher:webworker]
    cmd = chaussette --fd $(cvec_circus.sockets.webapp) chaussette.util.bench_app
    use_sockets = True

    [socket:webapp]
    host = 127.0.0.1
    port = 8888


plugin:NAME - as many sections as you want
==========================================
    **use**
        The fully qualified name that points to the plugin class.
    **anything else**
        Every other key found in the section is passed to the
        plugin constructor in the **config** mapping.

        You can use all the watcher options, since a plugin is
        started like a watcher.

Circus comes with a few pre-shipped :ref:`plugins <plugins>` but you can also extend them easily by :ref:`developing your own <develop_plugins>`.


env or env[:WATCHERS] - as many sections as you want
====================================================
    **anything**
        The name of an environment variable to assign value to.
        bash style environment substitutions are supported.
        for example, append /bin to `PATH` 'PATH = $PATH:/bin'

Section responsible for delivering environment variable to run processes.

Example:

.. code-block:: ini

    [watcher:worker1]
    cmd = ping 127.0.0.1

    [watcher:worker2]
    cmd = ping 127.0.0.1

    [env]
    CAKE = lie

The variable `CAKE` will propagated to all watchers defined in config file.

WATCHERS can be a comma separated list of watcher sections to apply this environment to.
if multiple env sections match a watcher, they will be combine in the order they appear in the configuration file.
later entries will take precedence.

Example:

.. code-block:: ini

    [watcher:worker1]
    cmd = ping 127.0.0.1

    [watcher:worker2]
    cmd = ping 127.0.0.1

    [env:worker1,worker2]
    PATH = /bin

    [env:worker1]
    PATH = $PATH

    [env:worker2]
    CAKE = lie

`worker1` will be run with PATH = $PATH (expanded from the environment cvec_circusd was run in)
`worker2` will be run with PATH = /bin and CAKE = lie

It's possible to use wildcards as well.

Example:

.. code-block:: ini

    [watcher:worker1]
    cmd = ping 127.0.0.1

    [watcher:worker2]
    cmd = ping 127.0.0.1

    [env:worker*]
    PATH = /bin


Both `worker1` and `worker2` will be run with PATH = /bin


Using environment variables
===========================

When writing your configuration file, you can use environment
variables defined in the *env* section or in *os.environ* itself.

You just have to use the *cvec_circus.env.* prefix.

Example:

.. code-block:: ini

    [watcher:worker1]
    cmd = $(cvec_circus.env.shell)

    [watcher:worker2]
    baz = $(cvec_circus.env.user)
    bar = $(cvec_circus.env.yeah)
    sup = $(cvec_circus.env.oh)

    [socket:socket1]
    port = $(cvec_circus.env.port)

    [plugin:plugin1]
    use = some.path
    parameter1 = $(cvec_circus.env.plugin_param)

    [env]
    yeah = boo

    [env:worker2]
    oh = ok

If a variable is defined in several places, the most specialized
value has precedence: a variable defined in *env:XXX* will override
a variable defined in *env*, which will override a variable
defined in *os.environ*.

environment substitutions can be used in any section of the configuration
in any section variable.


.. _formatting_cmd:

Formatting the commands and arguments with dynamic variables
============================================================

As you may have seen, it is possible to pass some information that are computed
dynamically when running the processes. Among other things, you can get the
worker id (WID) and all the options that are passed to the :class:`Process`.
Additionally, it is possible to access the options passed to the
:class:`Watcher` which instanciated the process.

.. note::

   The worker id is different from the process id. It's a unique value,
   starting at 1, which is only unique for the watcher.

For instance, if you want to access some variables that are contained in the
environment, you would need to do it with a setting like this::

    cmd = "make-me-a-coffee --sugar $(CIRCUS.ENV.SUGAR_AMOUNT)"

This works with both `cmd` and `args`.

**Important**:

- All variables are prefixed with `cvec_circus.`
- The replacement is case insensitive.

Stream configuration
====================

Simple stream class like `QueueStream` and `StdoutStream` don't have
specific attributes but some other stream class may have some:


FileStream
::::::::::

    **filename**
        The file path where log will be written.

    **time_format**
        The strftime format that will be used to prefix each time with a timestamp.
        By default they will be not prefixed.

        i.e: %Y-%m-%d %H:%M:%S

    **max_bytes**
        The max size of the log file before a new file is started.
        If not provided, the file is not rolled over.

    **backup_count**
        The number of log files that will be kept
        By default backup_count is null.


.. note::

    Rollover occurs whenever the current log file is nearly max_bytes in
    length. If backup_count is >= 1, the system will successively create
    new files with the same pathname as the base file, but with extensions
    ".1", ".2" etc. appended to it. For example, with a backup_count of 5
    and a base file name of "app.log", you would get "app.log",
    "app.log.1", "app.log.2", ... through to "app.log.5". The file being
    written to is always "app.log" - when it gets filled up, it is closed
    and renamed to "app.log.1", and if files "app.log.1", "app.log.2" etc.
    exist, then they are renamed to "app.log.2", "app.log.3" etc.
    respectively.

Example:

.. code-block:: ini

    [watcher:myprogram]
    cmd = python -m myapp.server

    stdout_stream.class = FileStream
    stdout_stream.filename = test.log

cvec_circus/config.py
import glob
import os
import signal
import warnings
from fnmatch import fnmatch
try:
    import resource
except ImportError:
    resource = None     # NOQA

import six

from cvec_circus import logger
from cvec_circus.py3compat import sort_by_field
from cvec_circus.util import (DEFAULT_ENDPOINT_DEALER, DEFAULT_ENDPOINT_SUB,
                         DEFAULT_ENDPOINT_MULTICAST, DEFAULT_ENDPOINT_STATS,
                         StrictConfigParser, replace_gnu_args, to_signum,
                         to_bool, papa)


def watcher_defaults():
    return {
        'name': '',
        'cmd': '',
        'args': '',
        'numprocesses': 1,
        'warmup_delay': 0,
        'executable': None,
        'working_dir': None,
        'shell': False,
        'uid': None,
        'gid': None,
        'send_hup': False,
        'stop_signal': signal.SIGTERM,
        'stop_children': False,
        'max_retry': 5,
        'graceful_timeout': 30,
        'rlimits': dict(),
        'stderr_stream': dict(),
        'stdout_stream': dict(),
        'priority': 0,
        'use_sockets': False,
        'singleton': False,
        'copy_env': False,
        'copy_path': False,
        'hooks': dict(),
        'respawn': True,
        'autostart': True,
        'use_papa': False}


class DefaultConfigParser(StrictConfigParser):

    def __init__(self, *args, **kw):
        StrictConfigParser.__init__(self, *args, **kw)
        self._env = dict(os.environ)

    def set_env(self, env):
        self._env = dict(env)

    def get(self, section, option, **kwargs):
        res = StrictConfigParser.get(self, section, option, **kwargs)
        return replace_gnu_args(res, env=self._env)

    def items(self, section, noreplace=False):
        items = StrictConfigParser.items(self, section)
        if noreplace:
            return items

        return [(key, replace_gnu_args(value, env=self._env))
                for key, value in items]

    def dget(self, section, option, default=None, type=str):
        if not self.has_option(section, option):
            return default

        value = self.get(section, option)

        if type is int:
            value = int(value)
        elif type is bool:
            value = to_bool(value)
        elif type is float:
            value = float(value)
        elif type is not str:
            raise NotImplementedError()

        return value


def rlimit_value(val):
    if resource is not None and (val is None or len(val) == 0):
        return resource.RLIM_INFINITY
    else:
        return int(val)


def read_config(config_path):
    cfg = DefaultConfigParser()
    with open(config_path) as f:
        if hasattr(cfg, 'read_file'):
            cfg.read_file(f)
        else:
            cfg.readfp(f)

    current_dir = os.path.dirname(config_path)

    # load included config files
    includes = []

    def _scan(filename, includes):
        if os.path.abspath(filename) != filename:
            filename = os.path.join(current_dir, filename)

        paths = glob.glob(filename)
        if paths == []:
            logger.warn('%r does not lead to any config. Make sure '
                        'include paths are relative to the main config '
                        'file' % filename)
        includes += paths

    for include_file in cfg.dget('cvec_circus', 'include', '').split():
        _scan(include_file, includes)

    for include_dir in cfg.dget('cvec_circus', 'include_dir', '').split():
        _scan(os.path.join(include_dir, '*.ini'), includes)

    logger.debug('Reading config files: %s' % includes)
    return cfg, [config_path] + cfg.read(includes)


def get_config(config_file):
    if not os.path.exists(config_file):
        raise IOError("the configuration file %r does not exist\n" %
                      config_file)

    cfg, cfg_files_read = read_config(config_file)
    dget = cfg.dget
    config = {}

    # reading the global environ first
    global_env = dict(os.environ.items())
    local_env = dict()

    # update environments with [env] section
    if 'env' in cfg.sections():
        local_env.update(dict(cfg.items('env')))
        global_env.update(local_env)

    # always set the cfg environment
    cfg.set_env(global_env)

    # main cvec_circus options
    config['check_delay'] = dget('cvec_circus', 'check_delay', 5., float)
    config['endpoint'] = dget('cvec_circus', 'endpoint', DEFAULT_ENDPOINT_DEALER)
    config['endpoint_owner'] = dget('cvec_circus', 'endpoint_owner', None, str)
    config['pubsub_endpoint'] = dget('cvec_circus', 'pubsub_endpoint',
                                     DEFAULT_ENDPOINT_SUB)
    config['multicast_endpoint'] = dget('cvec_circus', 'multicast_endpoint',
                                        DEFAULT_ENDPOINT_MULTICAST)
    config['stats_endpoint'] = dget('cvec_circus', 'stats_endpoint', None)
    config['statsd'] = dget('cvec_circus', 'statsd', False, bool)
    config['umask'] = dget('cvec_circus', 'umask', None)
    if config['umask']:
        config['umask'] = int(config['umask'], 8)

    if config['stats_endpoint'] is None:
        config['stats_endpoint'] = DEFAULT_ENDPOINT_STATS
    elif not config['statsd']:
        warnings.warn("You defined a stats_endpoint without "
                      "setting up statsd to True.",
                      DeprecationWarning)
        config['statsd'] = True

    config['warmup_delay'] = dget('cvec_circus', 'warmup_delay', 0, int)
    config['httpd'] = dget('cvec_circus', 'httpd', False, bool)
    config['httpd_host'] = dget('cvec_circus', 'httpd_host', 'localhost', str)
    config['httpd_port'] = dget('cvec_circus', 'httpd_port', 8080, int)
    config['debug'] = dget('cvec_circus', 'debug', False, bool)
    config['debug_gc'] = dget('cvec_circus', 'debug_gc', False, bool)
    config['pidfile'] = dget('cvec_circus', 'pidfile')
    config['loglevel'] = dget('cvec_circus', 'loglevel')
    config['logoutput'] = dget('cvec_circus', 'logoutput')
    config['loggerconfig'] = dget('cvec_circus', 'loggerconfig', None)
    config['fqdn_prefix'] = dget('cvec_circus', 'fqdn_prefix', None, str)
    config['papa_endpoint'] = dget('cvec_circus', 'fqdn_prefix', None, str)

    # Initialize watchers, plugins & sockets to manage
    watchers = []
    plugins = []
    sockets = []

    for section in cfg.sections():
        section_items = dict(cfg.items(section))
        if list(section_items.keys()) in [[], ['__name__']]:
            # Skip empty sections
            continue
        if section.startswith("socket:"):
            sock = section_items
            sock['name'] = section.split("socket:")[-1].lower()
            sock['so_reuseport'] = dget(section, "so_reuseport", False, bool)
            sock['replace'] = dget(section, "replace", False, bool)
            sockets.append(sock)

        if section.startswith("plugin:"):
            plugin = section_items
            plugin['name'] = section
            if 'priority' in plugin:
                plugin['priority'] = int(plugin['priority'])
            plugins.append(plugin)

        if section.startswith("watcher:"):
            watcher = watcher_defaults()
            watcher['name'] = section.split("watcher:", 1)[1]

            # create watcher options
            for opt, val in cfg.items(section, noreplace=True):
                if opt in ('cmd', 'args', 'working_dir', 'uid', 'gid'):
                    watcher[opt] = val
                elif opt == 'numprocesses':
                    watcher['numprocesses'] = dget(section, 'numprocesses', 1,
                                                   int)
                elif opt == 'warmup_delay':
                    watcher['warmup_delay'] = dget(section, 'warmup_delay', 0,
                                                   int)
                elif opt == 'executable':
                    watcher['executable'] = dget(section, 'executable', None,
                                                 str)
                # default bool to False
                elif opt in ('shell', 'send_hup', 'stop_children',
                             'close_child_stderr', 'use_sockets', 'singleton',
                             'copy_env', 'copy_path', 'close_child_stdout'):
                    watcher[opt] = dget(section, opt, False, bool)
                elif opt == 'stop_signal':
                    watcher['stop_signal'] = to_signum(val)
                elif opt == 'max_retry':
                    watcher['max_retry'] = dget(section, "max_retry", 5, int)
                elif opt == 'graceful_timeout':
                    watcher['graceful_timeout'] = dget(
                        section, "graceful_timeout", 30, int)
                elif opt.startswith('stderr_stream') or \
                        opt.startswith('stdout_stream'):
                    stream_name, stream_opt = opt.split(".", 1)
                    watcher[stream_name][stream_opt] = val
                elif opt.startswith('rlimit_'):
                    limit = opt[7:]
                    watcher['rlimits'][limit] = rlimit_value(val)
                elif opt == 'priority':
                    watcher['priority'] = dget(section, "priority", 0, int)
                elif opt == 'use_papa' and dget(section, 'use_papa', False,
                                                bool):
                    if papa:
                        watcher['use_papa'] = True
                    else:
                        warnings.warn("Config file says use_papa but the papa "
                                      "module is missing.",
                                      ImportWarning)
                elif opt.startswith('hooks.'):
                    hook_name = opt[len('hooks.'):]
                    val = [elmt.strip() for elmt in val.split(',', 1)]
                    if len(val) == 1:
                        val.append(False)
                    else:
                        val[1] = to_bool(val[1])

                    watcher['hooks'][hook_name] = val
                # default bool to True
                elif opt in ('check_flapping', 'respawn', 'autostart',
                             'close_child_stdin'):
                    watcher[opt] = dget(section, opt, True, bool)
                else:
                    # freeform
                    watcher[opt] = val

            if watcher['copy_env']:
                watcher['env'] = dict(global_env)
            else:
                watcher['env'] = dict(local_env)

            watchers.append(watcher)

    # making sure we return consistent lists
    sort_by_field(watchers)
    sort_by_field(plugins)
    sort_by_field(sockets)

    # Second pass to make sure env sections apply to all watchers.

    def _extend(target, source):
        for name, value in source:
            if name in target:
                continue
            target[name] = value

    def _expand_vars(target, key, env):
        if isinstance(target[key], six.string_types):
            target[key] = replace_gnu_args(target[key], env=env)
        elif isinstance(target[key], dict):
            for k in target[key].keys():
                _expand_vars(target[key], k, env)

    def _expand_section(section, env, exclude=None):
        if exclude is None:
            exclude = ('name', 'env')

        for option in section.keys():
            if option in exclude:
                continue
            _expand_vars(section, option, env)

    # build environment for watcher sections
    for section in cfg.sections():
        if section.startswith('env:'):
            section_elements = section.split("env:", 1)[1]
            watcher_patterns = [s.strip() for s in section_elements.split(',')]
            env_items = dict(cfg.items(section, noreplace=True))

            for pattern in watcher_patterns:
                match = [w for w in watchers if fnmatch(w['name'], pattern)]

                for watcher in match:
                    watcher['env'].update(env_items)

    # expand environment for watcher sections
    for watcher in watchers:
        env = dict(global_env)
        env.update(watcher['env'])
        _expand_section(watcher, env)

    config['watchers'] = watchers
    config['plugins'] = plugins
    config['sockets'] = sockets
    return config
